<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>specifyBSP function — specifyBSP • AlphaSimHlpR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="specifyBSP function — specifyBSP" />
<meta property="og:description" content="Function to manually specify a breeding scheme parameters (bsp) object in R, rather than using a control file.
Currently does not handle costs. For costs (for now), run specifyCosts() on this.
Ideally this is useful for programmatically varying breeding schemes." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AlphaSimHlpR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/AlphaSimHlpR_Manual.html">AlphaSimHlpR Manual</a>
    </li>
    <li>
      <a href="../articles/Using_AlphaSimHlpR.html">Using AlphaSimHlpR</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>specifyBSP function</h1>
    
    <div class="hidden name"><code>specifyBSP.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function to manually specify a breeding scheme parameters (bsp) object in R, rather than using a control file.
Currently does not handle costs. For costs (for now), run specifyCosts() on this.
Ideally this is useful for programmatically varying breeding schemes.</p>
    </div>

    <pre class="usage"><span class='fu'>specifyBSP</span><span class='op'>(</span>
  <span class='va'>schemeDF</span>,
  <span class='va'>nChr</span>,
  <span class='va'>effPopSize</span>,
  quickHaplo <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>segSites</span>,
  <span class='va'>nQTL</span>,
  <span class='va'>nSNP</span>,
  <span class='va'>genVar</span>,
  gxeVar <span class='op'>=</span> <span class='cn'>NULL</span>,
  gxyVar <span class='op'>=</span> <span class='cn'>NULL</span>,
  gxlVar <span class='op'>=</span> <span class='cn'>NULL</span>,
  gxyxlVar <span class='op'>=</span> <span class='cn'>NULL</span>,
  meanDD <span class='op'>=</span> <span class='fl'>0</span>,
  varDD <span class='op'>=</span> <span class='fl'>0</span>,
  relAA <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>stageToGenotype</span>,
  <span class='va'>nParents</span>,
  <span class='va'>nCrosses</span>,
  <span class='va'>nProgeny</span>,
  usePolycrossNursery <span class='op'>=</span> <span class='cn'>FALSE</span>,
  nSeeds <span class='op'>=</span> <span class='cn'>NULL</span>,
  useOptContrib <span class='op'>=</span> <span class='cn'>FALSE</span>,
  nCandOptCont <span class='op'>=</span> <span class='cn'>NULL</span>,
  targetEffPopSize <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>nClonesToNCRP</span>,
  phenoF1toStage1 <span class='op'>=</span> <span class='cn'>FALSE</span>,
  errVarPreStage1 <span class='op'>=</span> <span class='cn'>NULL</span>,
  useCurrentPhenoTrain <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>nCyclesToKeepRecords</span>,
  <span class='va'>selCritPipeAdv</span>,
  <span class='va'>selCritPopImprov</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>schemeDF</th>
      <td><p>data.frame columns: stageNames, nReps, nLocs, nChks, nEntries, entryToChkRatio, errVars</p><ul>
<li><p><code>stageNames</code>: Vector of number of number of entries in each stage</p></li>
<li><p><code>nReps</code>: Vector of number of reps used in each stage</p></li>
<li><p><code>nLocs</code>: Vector of number of locations used in each stage</p></li>
<li><p><code>nChks</code>: Vector of number of checks used in each stage. Checks are replicated the same as experimental entries</p></li>
<li><p><code>nEntries</code>: Vector of number of number of entries in each stage</p></li>
<li><p><code>entryToChkRatio</code>: How many entry plots do you have per check plot</p></li>
<li><p><code>errVars</code>: Vector of error variances estimated from historical data</p></li>
<li><p><code>trainingPopCycles</code>: How many cycles back to keep records from
 each stage in the training population. More cycles means bigger training
 population but also more distant from the selection candidates.
 Fewer cycles means the simulation runs faster</p></li>
</ul></td>
    </tr>
    <tr>
      <th>nChr</th>
      <td><p>integer number of chromosomes for the species</p></td>
    </tr>
    <tr>
      <th>effPopSize</th>
      <td><p>numeric historic effective population size for the species</p></td>
    </tr>
    <tr>
      <th>quickHaplo</th>
      <td><p>T/F, Whether to use the AlphaSimR "quickHaplo" feature. It is less realistic but greatly accelerates the creation of founders. Useful for quick tests.</p></td>
    </tr>
    <tr>
      <th>segSites</th>
      <td><p>integer number of sites segregating per chromosome</p></td>
    </tr>
    <tr>
      <th>nQTL</th>
      <td><p>integer number of loci affecting the trait per chromosome</p></td>
    </tr>
    <tr>
      <th>nSNP</th>
      <td><p>integer number of observed SNPs per chromosome</p></td>
    </tr>
    <tr>
      <th>genVar</th>
      <td><p>numeric genetic variance of the founders</p></td>
    </tr>
    <tr>
      <th>gxeVar</th>
      <td><p>numeric genotype by environment variance of the founders</p></td>
    </tr>
    <tr>
      <th>gxyVar</th>
      <td><p>genotype-by-year variance</p></td>
    </tr>
    <tr>
      <th>gxlVar</th>
      <td><p>genotype-by-location variance</p></td>
    </tr>
    <tr>
      <th>gxyxlVar</th>
      <td><p>genotype-by-location-by-year variance</p></td>
    </tr>
    <tr>
      <th>meanDD</th>
      <td><p>numeric mean dominance deviation. Set to zero for additive</p></td>
    </tr>
    <tr>
      <th>varDD</th>
      <td><p>numeric variance across loci of their dominance deviation</p></td>
    </tr>
    <tr>
      <th>relAA</th>
      <td><p>from AlphaSimR "the relative value of additive-by-additive variance compared to additive variance in a diploid organism with allele frequency 0.5"</p></td>
    </tr>
    <tr>
      <th>stageToGenotype</th>
      <td><p>At what stage do you want to genotype individuals?  In addition to the
named stages (SDN, CET, and PYT in this control file), the user can
specify F1, which will cause all nCrosses * nProgeny individuals to be
genotyped. If nothing is specified, the default will be to genotype all F1.</p></td>
    </tr>
    <tr>
      <th>nParents</th>
      <td><p>integer number of parents to cross</p></td>
    </tr>
    <tr>
      <th>nCrosses</th>
      <td><p>integer how many crosses to make</p></td>
    </tr>
    <tr>
      <th>nProgeny</th>
      <td><p>integer how many progeny per cross</p></td>
    </tr>
    <tr>
      <th>usePolycrossNursery</th>
      <td><p>T/F. Whether to use a polycross nursery. If it is used, nSeeds are made using completely random mating</p></td>
    </tr>
    <tr>
      <th>nSeeds</th>
      <td><p>Parameter to determine the number of seeds only if usePolycrossNursery TRUE</p></td>
    </tr>
    <tr>
      <th>useOptContrib</th>
      <td><p>logical whether to use optimal contributions</p></td>
    </tr>
    <tr>
      <th>nCandOptCont</th>
      <td><p>integer how many candidates to consider for opt contrib</p></td>
    </tr>
    <tr>
      <th>targetEffPopSize</th>
      <td><p>numeric target effective population size for OC</p></td>
    </tr>
    <tr>
      <th>nClonesToNCRP</th>
      <td><p>Number of clones that will be sent to NCRP for potential variety release</p></td>
    </tr>
    <tr>
      <th>phenoF1toStage1</th>
      <td><p>T/F. Parameters to control how seeds are moved from F1 to the Stage 1 trial</p></td>
    </tr>
    <tr>
      <th>errVarPreStage1</th>
      <td><p>Parameters to control how seeds are moved from F1 to the Stage 1 trial</p></td>
    </tr>
    <tr>
      <th>useCurrentPhenoTrain</th>
      <td><p>T/F. Are the current year phenotypes available for model training. when parents are chosen to send to the crossing nursery?</p></td>
    </tr>
    <tr>
      <th>nCyclesToKeepRecords</th>
      <td><p>integer eliminate data on cycles above this num</p></td>
    </tr>
    <tr>
      <th>selCritPipeAdv</th>
      <td><p>function used to determine selection criterion for pipe</p></td>
    </tr>
    <tr>
      <th>selCritPopImprov</th>
      <td><p>function used to determine sel crit for pop improv</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a named list of of the parameters to specify a breeding scheme simulation</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>All arguments are exactly as specified in the control files. Main exception is schemeDF, which is just a tibble() or data.frame version of the set of bsp arguments which are vectors (giving values for each breeding stage). Columns must have names exactly as in the corresponding arguments in control file: stageNames, nReps, nLocs, nChks, nEntries, entryToChkRatio, errVars</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


